### PR TITLE
[MPS] Add lu_factor

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99269

### <samp>🤖 Generated by Copilot at d75cde1</samp>

Added MPS support and autograd formulas for LU factorization of tensors. Implemented the `linalg_lu_factor` and `linalg_lu_factor.out` functions for the MPS backend in `LinearAlgebra.mm` and added tests in `test_mps.py`. Added the corresponding dispatch entries in `native_functions.yaml` and the backward and forward formulas in `derivatives.yaml`.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (qqaatw) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:136]: For `A_t.dim() > 3` cases we should clone `A_`; otherwise the input will be modified in-place. (Removing the above block:
```python
if (!A_.is_contiguous()) {
    A_ = A_.clone(at::MemoryFormat::Contiguous);
  }
```
)

File: aten/src/ATen/native/native_functions.yaml
Comment (lezcano) [inline aten/src/ATen/native/native_functions.yaml:13159]: Could you implement `linalg_lu_factor_ex` rather than this one? That way you wouldn't need to add any new backward rule, and all the other goodies that are implemented for `linalg.lu_factor`  will also extend to MPS.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (lezcano) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:130]: Why not just a tensor with `batchSize` elements in the case of `status_tensors`? Same same for `pivots_list`.

File: test/test_mps.py
Comment (lezcano) [inline test/test_mps.py:6946]: This doesn't test the `pivot` kwarg.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (lezcano) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:108]: What happens if `pivot` is `false`?  Same below.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (lezcano) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:115]: `flatten` may copy. Do you really need to copy here? You can see how this is done in CPU/CUDA for inspiration on how to not perform unnecessary copies.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (lezcano) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:118]: Do you really want to do this when you're copying `A` into `LU` down below?
I think you want to:
1. Get an `LU` that always has the correct stride
2. Unconditionally copy `A` into `LU`.

You can see who this is done in the structured kernel path.

**Review by lezcano — COMMENTED** [review]
Left a few comments.
I'll let the MPS guys review the main implementation though.

File: aten/src/ATen/native/native_functions.yaml
Comment (qqaatw) [reply aten/src/ATen/native/native_functions.yaml:13159]: The reason I implement `linalg_lu_factor` is because `linalg_lu_factor_ex` has to return an `info` tensor, which is not applicable to MPS. (the info tensor is computed by LAPACK on cpu, for example).

Comment (qqaatw) [reply aten/src/ATen/native/native_functions.yaml:13159]: I guess I cannot directly return an undefined Tensor `Tensor()` as the `info` since there are some post check logics to the `info` tensor.

File: test/test_mps.py
Comment (qqaatw) [reply test/test_mps.py:6946]: Thanks for pointing it out! I planned to remove this test and rely on the OpInfo test only after the unimplemented backward pass is completed. I will add a TODO here.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (lezcano) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm:126]: I do not know enough about MPS, but perhaps @kulinseth can comment on what's the best way to do this. It'd be good if `pivots` could be a `Tensor`, because then we could implement the `_ex` version of the function, which would be the preferred way of implementing all this.

File: test/test_mps.py
Comment (lezcano) [reply test/test_mps.py:6969]: Then, rather than adding this test, enable the OpInfo for this operation and xfail the gradient tests on mps.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (lezcano) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:100]: nit. `linalg.lu_factor`.

**Review by kulinseth — CHANGES_REQUESTED** [review]
I agree with comments earlier @lezcano ..

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm:126]: Sorry, should it be initialized with `[ndArray buffer]` or `[arrayView buffer]`? I guess the latter is what you were suggesting.

With pivot matrix initialized with `[arrayView buffer]` that uses `MPSAliasingStrategyShallNotAlias`, the pivots output remains the same as the pivots before LU decomposition, which looks like the array view is not writable with this strategy? On the other hand, if I specify `MPSAliasingStrategyShallAlias`, the output is correct with unbatched inputs.

I also tried initializing with `[ndArray buffer]`, the outputs were incorrect if the inputs were batched.

Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm:126]: @kulinseth I can provide the code if you need repro.

**Review by kulinseth — APPROVED** [review]
Thanks, looks good.

**Comment by lezcano** [comment]
Most of the points above are still relevant tho. There are a number of copies that could be elided just by making the code a tad more general. Also, we should reuse the testing we already have for these functions.

**Comment by qqaatw** [comment]
> Most of the points above are still relevant tho. There are a number of copies that could be elided just by making the code a tad more general. Also, we should reuse the testing we already have for these functions.

Sure, I will revisit the code.

File: test/test_mps.py
Comment (qqaatw) [reply test/test_mps.py]: Done.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm:120]: We don't need copy here but end up need `A_` to be contiguous before feeding it into the kernel. Can you please point out where it is done in CPU/CUDA?

Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm:120]: Hi @lezcano, can you give me some pointer here on how the cpu/cuda avoid unnecessary copies?

Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm]: Used view instead

**Review by lezcano — APPROVED** [review]
There is a a minor correctness issue. That being said, the PR looks mostly right. @kulinseth already approved, so feel free to merge after either fixing the issue properly or simply reverting to the previous flatten (not entirely efficient, but at least correct).

**Pull request closed without merging**
